[Design] 통계화면 구현

futuremirae opened the pull request:
### 📌 관련 이슈 번호 ex) #이슈번호
- #5 

# 📘 작업 유형
 - [x] 신규 기능 추가

# 📙 작업 내역 (구현 내용 및 작업 내역을 기재합니다.)
 - 통계 View 생성 
 
<img width="492" alt="스크린샷 2024-11-23 오후 6 50 10" src="https://github.com/user-attachments/assets/df991e85-cd28-4581-a12b-24b18b99c90f">

- 상단의 표를 보면```지난 30일``` 이 부분이 너비가 안맞는데,, 해당 pr 머지되면 수정하도록 하겠습니다. 빨리 검사 받고 싶어서 올렸습니다~~!!

### 📝 특이 사항 (Optional)
- 사실 이모티콘, 색상 추가등 아직 추가해야할 요소들이 많지만 먼저 리뷰를 받고 수정한 후에 추가하고 싶어 그린 뷰만 우선적으로 pr을 올렸습니다. !!

- 이번 작업에서는 view들이 많이 쪼개져 있습니다! 

```
  struct State: Equatable {
    var userName: String
    var totalCount: Int
    var sequenceCount: Int
    var monthCount: Int
    var sadnessCount: Int
    var happinessCount: Int
    var surpriseCount: Int
    var fearCount: Int
    var disgustCount: Int
    var neutralCount: Int
    var noneCount: Int
    var emotion: String
    var reply: String
  }
```
- 현재 뷰모델의 state에는 이런식으로 정의가 되어있는데요, 
- 처음 생각을 했을때 repo에서 넘어와야할 정보는 크게
    - Username, 
    - 감정(최근 30일 감정 8개)
    - 쓴 일기의 갯수
    - 잼미니의 답장 
- 위와 같이 4개로 생각했습니다. state에는 엄청 여러개 만들어져 있는데, View에서는 개별적인  property로 정의 되어 있기 때문에 이렇게 설정을 해두었습니다. 

### 궁금한거! 
- Viewmodel에서 Action을 뭐로 해야하지 생각을 해보았는데
  -  통계화면에서는 처음에 데이터(사용자 이름, 일기 갯수, 감정 통계..)만  가져오면 되고 이후에는 별다은 action이 없는데,, 만들어야 하는건가?? 라는 생각이 들었습니다..! 맞을까요?! 
  - 그렇다면 state로 관찰을 계속 할 필요가 있는걸까요..?

futuremirae committed:
[Feat] -  #5 DefaultRepository 생성

- ViewController을 위한 Mock 객체

futuremirae committed:
[Feat] - #5 ReportRepository 생성

- ViewController을 위한 Mock 객체

futuremirae committed:
[Feat] #5 ReportUseCase 생성

- ViewController을 위한 Mock 객체
- 추후 비지니스 로직 fix 필요

futuremirae committed:
[Feat] - #5 ReportCoordinator

- ViewController을 위한 Mock 객체
- 추후 비지니스 로직 fix 필요

futuremirae committed:
[Feat] - #5 ReportViewController생성

futuremirae committed:
[Feat] - #5 ReportViewModel생성

- ViewController을 위한 Mock객체
- 추후 수정필요

futuremirae committed:
[Style] - #5 View작업을 위한 기본 코드 수정

- View작업을 위해서 필요한 Controller을 위해 Coordinator과 viewModel 부분 수정
- 특히 ViewModel부분에서 UseCase가 있을 경우 Dicontainer을 통해 의존성 주입이 필요하기 떄문에 해당 부분 TODO 처리

futuremirae committed:
[Design] - #5 ReportCountView 파일 생성 및 구현

futuremirae committed:
[Design] - #5 ReportView파일 생성 및 구현

- 상단의 레이블과 통계표 구현

futuremirae committed:
[Feat] - #5 ReportViewController에 View추가

- 구현한 상단 뷰 Controller에 추가
- 필요없는 주석 제거
- 하단뷰 추가 필요

futuremirae committed:
[Design] GraphView 파일 생성 완료

futuremirae committed:
[Design] - #5 BarView 구현

- 그래프 안에 들어가는 막대

futuremirae committed:
[Design] - #5 GraphView 구현

- BarView들을 스택뷰로 묶어 막대그래프로 구현

futuremirae committed:
[Chore] - #5 공백 제거

futuremirae committed:
[Design] - #5 graphView 레이아웃

futuremirae committed:
[Feat] - #5 Chart생성

- 막대그래프의 높이와 감정에 따른 생상을 결정합니다.

futuremirae committed:
[Design] #5 GraphView 레이아웃 변경

- 스택뷰 레이아웃 오류 해결

futuremirae committed:
[Design] #5 ReportView 레이아웃 변경

futuremirae committed:
[Design] #5 하임이의 답장 뷰 구현

futuremirae committed:
[Feat] -  #5   AssetCatalog에 disgust아이콘 추가

futuremirae committed:
[Feat]  - #5  HeimEmoji class 생성

- 기본 감정 이코티콘 생성하는 class 생성
- static function으로 기본 감정 이코티콘 반환 가능

futuremirae committed:
[Design] - #5 레이아웃 수정 및 TODO 추가

futuremirae committed:
[Design] - #5 ScrollView 추가 및 레이아웃 수정

futuremirae committed:
[Style] - #5 주석 제거

futuremirae committed:
[Style] - #5 주석제거

futuremirae committed:
[Chore] - #5 주석 추가

futuremirae committed:
[Feat] - #5 ViewModel State 속성추가

futuremirae committed:
[Feat] - #5 parameter 수정

Hansangjin98 commented on Heim/Application/Application/Source/SceneDelegate.swift:
```diff
@@ -14,13 +14,14 @@ import NetworkModule
 import UIKit
 
 final class SceneDelegate: UIResponder, UIWindowSceneDelegate {
```
PR 올리기 전에 SceneDelegate는 항상 원상복구 부탁드려요!

Hansangjin98 commented on Heim/Presentation/Presentation/Report/Model/Chart.swift:
```diff
@@ -0,0 +1,20 @@
+//  Created by 김미래 on 11/22/24.
+//
+
+import UIKit
+// 막대그래프의 막대의 비율을 계산하는 모델, value = 해당 감정의 갯수/totalCount
+struct Chart {
```
Model은 UI와 분리된 독립적인 요소인데 UIKit에 의존해도 괜찮을까요?

ParkSeongGeun replied:
UIColor 속성을 사용하기 위해 UIKit을 사용하신 것 같은데, 이전 저희 스프린트 때도 진행했던 사항 중에 위 상진님 댓글처럼 Model은 UI와 분리된 독립적인 요소라는 설명이 있었던 것 같습니다.
- Chart에서 사용하는 색 정보를 다른 곳에서 정의하도록 하는 방안 등 분리에는 여러 방향이 존재할 거 같아요!

futuremirae replied:
```swift
struct Chart {
  // MARK: Properties
  var value: Double
  var color: UIColor

  // MARK: - Initializer
  init(value: Double, color: UIColor) {
    self.value = value
    self.color = color
  }
}
```
- 기존 코드가 Model에 존재함에도 불구하고 uikit을 import 해서 사용하고 있었습니다...

```swift
enum HeimEmotion {
  case sadness
  case happiness
  case angry
  case surprise
  case fear
  case disgust
  case neutral
}
struct Chart {
  // MARK: Properties
  var value: Double
  var emotion: HeimEmotion

  // MARK: - Initializer
  init(value: Double, emotion: HeimEmotion) {
    self.value = value
    self.emotion = emotion
  }
}
```
- 수정후에는,,,  color을 받는 것이 아니라 emtion을  받도록 하였습니다! 
- 해당 Chart는 ---> BarView가 생성될때 init의 인자로 들어갑니다! 
   - value를 통해 (막대의 높이)비율이 결정되고, emotion에 따라 색이 결정됩니다. 

```swift
func setBarColor(emotion: HeimEmotion) -> UIColor {
    switch emotion {
    case .sadness:
      return .red
    case .happiness:
      return .blue
    case .angry:
      return .white
    case .surprise:
      return .yellow
    case .fear:
      return .black
    case .disgust:
      return .orange
    case .neutral:
      return .brown
    }
  }
```
- BarView의 내부 함수에는 setBarColor()가 존재합니다. 이를 통해 감정에 따른 UiColor을 반환합니다! 
- 수정하고 나니 코드가 더 길어진거 같네요..

Hansangjin98 commented on Heim/Presentation/Presentation/Report/Model/HeimEmoji.swift:
```diff
@@ -0,0 +1,52 @@
+    case disgust
+    case neutral
+    case none
+  }
```
이미 동일하게 Emotion 엔티티가 정의되어 있어서 Emotion 엔티티를 써도 될 것 같긴 한데, Domain에 의존하지 않는 독립적인 View를 구성하기 위해서라면 지금처럼 재정의를 해도 괜찮겠네용

Hansangjin98 commented on Heim/Presentation/Presentation/Report/Model/HeimEmoji.swift:
```diff
@@ -0,0 +1,52 @@
+//
+
+import UIKit
+
+class HeimEmoji: UIImageView {
+  // MARK: Properties
```
final 빠뜨리신 것 같습니다!

해당 클래스는 Model일까요 View일까요? 목적을 고려해 역할을 다시 한번 생각해보시면 좋을 것 같습니다!

futuremirae replied:
뷰라고 생각합니다!! 
View 폴더로 옮기는 게 좋을 것 같아요!

Hansangjin98 commented on Heim/Presentation/Presentation/Report/View/BarView.swift:
```diff
@@ -0,0 +1,48 @@
+
+// 그래프의 막대 부분
+final class BarView: UIView {
+  let contentView: UIView = {
+    let view = UIView()
+    return view
+  }()
+
+  let bar: UIView = {
+    let view = UIView()
+    return view
+  }()
```
contentView와 bar는 정말 필요한 프로퍼티일까요?

futuremirae replied:
네  필요한 프로퍼티 입니다! 
contentView는 원래의 막대 전체 높이입니다. 
bar는( 해당 감정 갯수 / 전체 갯수 )의 비율값으로 들어가 contentView 위에 해당 비율만큼 그려지게 됩니다! 
위와 같이 만들어진 막대는 GraphView의 스택뷰에 차례대로 추가가되는데,, 
bar만 넣게 되면 높이가 그냥 다 일정하게 나옵니다..  

![스크린샷 2024-11-25 오후 12 23 33](https://github.com/user-attachments/assets/f417534d-dc0e-42db-8bd6-cbcb185393b9)

그래서 contentView를 추가해서 넣어줬습니다!

Hansangjin98 commented on Heim/Presentation/Presentation/Report/View/GraphView.swift:
```diff
@@ -0,0 +1,90 @@
+      $0.top.equalTo(graphStackView.snp.bottom).offset(LayoutConstants.defaultPadding)
+      $0.leading.trailing.equalTo(graphStackView)
+      $0.height.equalTo(self.snp.height).multipliedBy(0.2)
+    }
```
SnapKit을 사용할 때 self의 snp에 접근하는 경우는 사실상 없다고 생각하셔도 무방할 것 같아요!
`$0.height.equalToSuperview().multipliedBy(...)`

Hansangjin98 commented on Heim/Presentation/Presentation/Report/View/ReportViewController.swift:
```diff
@@ -0,0 +1,76 @@
+      return scrollView
+    }()
+
+  private let backgroundImageView: UIImageView = {
+    let imageView = UIImageView()
+    imageView.image = .background
+    imageView.contentMode = .scaleAspectFill
+    return imageView
+  }()
```
backgroundImageView는 이미 BaseViewController에서 깔아주고 있어서 불필요하겠네요!

futuremirae replied:
하하 백그라운드 이미지가 계속 안보여서 ,,, 저렇게 한건데 
super.setupLayoutConstraints()를 안써서 그런거 있죠,,,,,
흑 수정햇습니다 ㅜㅜㅜ

Hansangjin98 commented on Heim/Presentation/Presentation/Report/View/ReportViewController.swift:
```diff
@@ -0,0 +1,76 @@
+  override func viewDidAppear(_ animated: Bool) {
+    super.viewDidAppear(animated)
+    coordinator?.didFinish()
+  }
```
아마 Appear가 아니라 Disappear 시점에 `coordinator?.didFinish()`를 호출하려고 하신 것 같은데, 음성 녹음 버튼을 Tap 해서 RecordView가 present 되어 Disappear가 호출되는 경우도 있을 거라 coordinator 연결이 끊기지 않도록 didFinish 내부 로직을 수정해야겠네요!

Hansangjin98 commented on Heim/Presentation/Presentation/Report/ViewModel/ReportViewModel.swift:
```diff
@@ -0,0 +1,67 @@
+    case fetchData
+  }
+
+  struct State: Equatable {
+    var userName: String
+    var totalCount: Int
+    var sequenceCount: Int
+    var monthCount: Int
+    var sadnessCount: Int
+    var happinessCount: Int
+    var surpriseCount: Int
+    var fearCount: Int
+    var disgustCount: Int
+    var neutralCount: Int
+    var noneCount: Int
+    var emotion: String
+    var reply: String
+  }
```
각 감정에 대한 count는 별도의 Action이나 시점에 Update 되는 게 아니라 한번에 되기 때문에 묶어서 관리해도 좋을 것 같습니다. (휴먼 에러 방지)

감정 종류 중 none은 '분류되지 않는 경우'에 부여되는 결과(오류)이기 때문에 감정에 속하면 안될 것 같고, angry 감정이 빠졌네요

futuremirae replied:
```swift
  struct State: Equatable { // 해당 뷰모델 클래스 내부에 추가 !
    var userName: String
    var totalCount: Int
    var sequenceCount: Int
    var monthCount: Int
    var emotionCount: EmotionCount
    var emotion: String
    var reply: String
  }
  
  init() {
    self.state = State(userName: "미래",
                       totalCount: 0,
                       sequenceCount: 0,
                       monthCount: 0,
                       emotionCount: EmotionCount(sadnessCount: 0, // 요렇게 묶는 걸 말씀하신거가 맞을까요?!
                                                  happinessCount: 0,
                                                  surpriseCount: 0,
                                                  fearCount: 0,
                                                  disgustCount: 0,
                                                  neutralCount: 0,
                                                  angryCount: 0),
                       emotion: "",
                       reply: "답장이 도착하지 않았어요!")
  }
```
 
- 이렇게 EmotionCount로 묶는 걸 말씀하신 걸까요?!

Hansangjin98 reviewed (commented):
화면 구현에 대한 숙련도가 점차 늘어가시는 게 눈에 보이네요!
어떻게 그려나갈지 고민을 많이 하신 게 느껴지고, 그동안 조언받거나 공유받았던 내용들도 반영하고자 노력하신 게 보입니다 👍

---

Action과 State에 대해 고민이 있으신 것 같은데, 우선 State에 대해 생각을 해보시면 편할 것 같습니다.
State에는 View의 상태를 표현하는 데이터, View에서 변경 사항을 모니터링해야 하는 데이터가 포함되게 됩니다.
통계 화면에 나타나는 데이터들은 비즈니스 로직을 걸쳐 받아와야 하는 데이터들이죠? 바로 계산되는 데이터가 아니기 때문에 초기값과 이후 값에 변동이 생기게 됩니다. 변경 사항을 모니터링해야 하는 사항들이죠.
저희는 변경 사항 모니터링을 Combine을 활용해 바인딩하는 방식으로 처리를 하고 있으니 State에 포함되는 것은 자연스러워 보입니다.

State의 값은 Action을 통해 변경을 이루게 됩니다.
어떠한 경우에 변경이 이뤄지는지를 고려해서 Action을 설계하면 되는데, 현재 통계 화면의 경우 ViewDidLoad 시점에 모든 데이터를 불러오면 끝이기 때문에 미래님께서 전체 데이터를 불러오는 1개의 Action을 정의한 것은 자연스러워 보입니다.

---

아직 완료된 작업이 아니다 보니 기본적인 사항들에 미스가 많이 있는 것 같습니다.
함수/프로퍼티의 접근 제어자, 네이밍, 선언 타입 등등 전체적으로 코드 검토가 한번 필요할 것 같고, 레이아웃도 빠뜨리거나 중첩된 사항이 없는지 한번 확인 부탁드리겠습니다!
코멘트로 다 남기기에는 애매해서 혹시 한번 검토 이후에도 놓치시는 게 생기시면 말씀드리겠습니다!

futuremirae committed:
[Chore] -  #5  Scenedelegate 원래 코드로 복구

ParkSeongGeun commented on Heim/Presentation/Presentation/Report/View/GraphView.swift:
```diff
@@ -0,0 +1,90 @@
+final class GraphView: UIView {
+  // TODO: mock데이터 제거 및 color 추가
+  let emotionCharts: [Chart] =
+  [Chart(value: 0.1, color: .red),
```
이를 개산하는 로직 부분은 아직 미구현인 것이죠?!

futuremirae replied:
네 이 부분은 로직을 구현해야합니다!

ParkSeongGeun commented on Heim/Presentation/Presentation/Report/View/ReportView.swift:
```diff
@@ -0,0 +1,144 @@
+
+  // MARK: - Properties
+  var userName = "성근"
+  var emotion = "슬픔"
```
😭

ParkSeongGeun commented on Heim/Presentation/Presentation/Report/View/ReportView.swift:
```diff
@@ -0,0 +1,144 @@
+    let label = UILabel()
+    label.font = UIFont.boldFont(ofSize: LayoutConstants.titleThree)
+    label.numberOfLines = 0
+    label.text = """
```
현재는 출력 목적으로 이와 같이 작성해주신 것 같은데, 초기는 아무런 설정 없이 view가 load되고 감정 정보를 불러온 상태에서 label.text를 초기화 해주는 게 더 좋지 않을까요?

futuremirae replied:
```swift
private lazy var emotionLabel: UILabel = {
    let label = UILabel()
    label.font = UIFont.boldFont(ofSize: LayoutConstants.titleThree)
    label.numberOfLines = 0
    label.text = """
    지난 30일간 \(userName)님께서 
    가장 많이 느끼신 감정은 \(emotion)이군요
    """
    label.textColor = .white
    label.textAlignment = .center
    return label
  }()
```
현재는 이렇게 구현되어 있습니다. 
userName과 emotion은 처음 화면이 그려질 때 ViewModel에서 가져올 것 같아요!

ParkSeongGeun commented on Heim/Presentation/Presentation/Report/View/ReportView.swift:
```diff
@@ -0,0 +1,144 @@
+  }()
+
+  let replyTextView: CommonTextAreaView = {
+    let textAreaView = CommonTextAreaView()
```
CommonTextAreaView = CommonTextAreaView()로 해도 되는 거 같습니다!

ParkSeongGeun commented on Heim/Presentation/Presentation/Report/View/ReportViewController.swift:
```diff
@@ -0,0 +1,76 @@
+      let scrollView = UIScrollView()
+      scrollView.showsVerticalScrollIndicator = false
+      return scrollView
+    }()
```
혹시 못 보셨을까봐 남겨요! (한칸 띄워져있습니다.)

ParkSeongGeun reviewed (commented):
고생 많으셨습니다 미래님!
- UI를 영역별로 분리한 뒤, 작업하시는 실력이 많이 성장하신 것 같아 보고 배웁니다.

궁금한 점은 아래와 같습니다.
- 현재 이전의 데이터를 불러오기 위해 ReportRepository를 만드셨는데, 이와 같은 새로운 Repository를 만들 필요가 있을까 생각이 드네요.
    - 새로운 UseCase를 만드는 것은 타당하다고 생각합니다. (결국 Report를 만들기 위한 새로운 UseCase는 맞으니)
    - 다만, 이는 결국 DataStorageModule을 이용하여 필요한 작업을 수행하는 것이고, 기존 제작한 DiaryRepository를 활용하는 것이니 새로운 Repository를 정의하여 이를 이용할 이유가 있을까? 라는 생각이 들었습니다.

futuremirae committed:
[Refactor] - #5  폴더 변경 및 접근제어자 추가

- 해당 파일은 View이기 떄문에 View 폴더로 이동
- 접근제어자 final 추가

futuremirae committed:
[Refactor] - #5 LayoutConstants 추가 및 불필요한 Property 제거

futuremirae committed:
[Style] - #5 주석 제거 및 coordinator.didfinish() 호출 시점 변경

futuremirae committed:
[Refactor] - #5 리뷰 반영

- HeimEmotion 감정 7가지로 수정
- Chart Model이 uikit을 import 하지 않도록 수정
- ReportViewModel에서 EmotionCount 한번에 묶이도록 수정

Hansangjin98 reviewed (approved):
고생하셨습니다~!

futuremirae committed:
[Refacor] 리뷰반영

- 주석 제거
- 접근제어자 추가

ParkSeongGeun reviewed (approved):
UI 확인 후 Approve 하겠습니다.
- Repository / UseCase / Coordinator 는 추후 수정이 필요할 수도 있겠네요 :)

futuremirae committed:
Merge branch 'develop' into Design/Report/5
